server/PathServices/SignalRRealtimeDataRepository.cs

```diff
@@ -41,12 +43,36 @@ public SignalRRealtimeDataRepository(IPathDataRepository pathDataRepository, IPa
         /// <returns>A collection of arriving trains.</returns>
         public Task<IEnumerable<RealtimeData>> GetRealtimeData(Station station)
         {
-            return Task.FromResult(this.GetRealtimeData(station, RouteDirection.ToNY).Union(this.GetRealtimeData(station, RouteDirection.ToNJ)).Where(data => data.DataExpiration > DateTime.UtcNow));
+            var allData = this.GetRealtimeData(station, RouteDirection.ToNY).Union(this.GetRealtimeData(station, RouteDirection.ToNJ));
+            var freshData = allData.Where(dataPoint => dataPoint.DataExpiration > DateTime.UtcNow);
+            if (allData.Count() != freshData.Count())
+            {
+                var staledData = allData.Except(freshData);
+                foreach (var staledDataPoint in staledData)
+                    Log.Logger.Here().Warning("Staled data detected for S:{station} R:{route} with timestamp {updatedDataLastUpdated}, force reconnect maybe needed", station, staledDataPoint.Route.DisplayName, staledDataPoint.LastUpdated);
+
+                Log.Logger.Here().Information("Recreating SignalR hubs following staled data detection...");
+                Task.Run(this.CreateHubConnections).Wait();
+            }
+            return Task.FromResult(freshData);
         }
 
         private IEnumerable<RealtimeData> GetRealtimeData(Station station, RouteDirection direction)
         {
-            return this.realtimeData.GetValueOrDefault((station, direction), new List<RealtimeData>());
+            Log.Logger.Here().Debug("Getting realtime data for {station}-{direction}...", station, direction);
+            var startTimestamp = DateTime.UtcNow;
```

Comment:
You could use StopWatch to get the runtime rather than subtracting two DateTimes.